File: services/sabwa-engine/src/routes/sessions.rs
//! `/sessions` — pairing, lifecycle and status for linked WhatsApp accounts.
//!
//! Mirrors the server actions from SABWA_PLAN.md §13 (`pairSession`,
//! `logoutSession`, `renameSession`, `listSessions`) plus a status endpoint
//! used by the browser as a polling fallback when SSE is unavailable.

use axum::{
    extract::{Path, Query, State},
    http::HeaderMap,
    routing::get,
    Json, Router,
};
use serde::{Deserialize, Serialize};

use crate::audit::{self, AuditEntry};
use crate::error::AppError;
use crate::state::AppState;

/// Build the `/sessions` sub-router.
pub fn router() -> Router<AppState> {
    Router::new()
        .route("/", get(list_sessions).post(create_session))
        .route(
            "/:id",
            get(get_session).delete(delete_session).patch(update_session),
        )
        .route("/:id/status", get(get_session_status))
}

// ---------- DTOs ----------

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CreateSessionRequest {
    pub project_id: String,
    pub user_id: String,
    /// `qr` or `code`.
    pub pair_method: String,
    pub phone_e164: Option<String>,
}

#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct CreateSessionResponse {
    pub session_id: String,
    pub status: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub qr: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub pair_code: Option<String>,
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ListSessionsQuery {
    pub project_id: String,
}

#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct SessionSummary {
    pub session_id: String,
    pub project_id: String,
    pub user_id: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub phone_e164: Option<String>,
    pub status: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub push_name: Option<String>,
}

#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct ListSessionsResponse {
    pub sessions: Vec<SessionSummary>,
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct UpdateSessionRequest {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub label: Option<String>,
    /// `safe` | `normal` | `aggressive`.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub rate_limit_profile: Option<String>,
}

#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct SessionStatusResponse {
    pub session_id: String,
    pub status: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub qr: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub pair_code: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub last_connected_at: Option<chrono::DateTime<chrono::Utc>>,
}

#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct DeletedResponse {
    pub session_id: String,
    pub deleted: bool,
}

#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct UpdatedResponse {
    pub session_id: String,
    pub updated: bool,
}

// ---------- Handlers ----------

async fn create_session(
    State(state): State<AppState>,
    headers: HeaderMap,
    Json(body): Json<CreateSessionRequest>,
) -> Result<Json<CreateSessionResponse>, AppError> {
    tracing::info!(
        project_id = %body.project_id,
        user_id = %body.user_id,
        pair_method = %body.pair_method,
        "sessions: create"
    );

    let (actor_ip, user_agent) = audit::extract_context(&headers);

    // Persist a `pending` row up front so:
    //   1. `listSessions` reflects the new account immediately.
    //   2. The connected-event persister in `wa::baileys::persist_auth_state`
    //      can resolve the session by ObjectId and flip status → connected.
    // Previously this route synthesized a `sess_<uuid>` id and skipped the
    // insert entirely, which left the auth-state persister logging
    // "session_id is not a valid ObjectId — skipping" and the overview UI
    // permanently stuck at "0 accounts".
    let project_oid = bson::oid::ObjectId::parse_str(&body.project_id)
        .map_err(|_| AppError::BadRequest(format!("invalid projectId: {}", body.project_id)))?;
    let user_oid = bson::oid::ObjectId::parse_str(&body.user_id)
        .map_err(|_| AppError::BadRequest(format!("invalid userId: {}", body.user_id)))?;
    let pair_method_db = match body.pair_method.as_str() {
        "code" => crate::db::sessions::PairMethod::Code,
        "qr" => crate::db::sessions::PairMethod::Qr,
        other => {
            return Err(AppError::BadRequest(format!(
                "pairMethod must be 'qr' or 'code', got '{other}'"
            )))
        }
    };
    let now = chrono::Utc::now();
    let pending_row = crate::db::sessions::SabwaSession {
        id: None,
        project_id: project_oid,
        user_id: user_oid,
        phone_e164: body.phone_e164.clone(),
        push_name: None,
        profile_pic_url: None,
        status: crate::db::sessions::SessionStatus::Pending,
        pair_method: pair_method_db,
        auth_state: None,
        device_meta: None,
        last_connected_at: None,
        last_seen_at: None,
        worker_node_id: None,
        ban_signals: Vec::new(),
        rate_limit_profile: crate::db::sessions::RateProfile::Normal,
        created_at: now,
        updated_at: now,
    };
    let inserted_id = crate::db::sessions::SessionsRepo::new(&state.db)
        .insert(&pending_row)
        .await?;
    let session_id = inserted_id.to_hex();

    let method = match body.pair_method.as_str() {
        "code" => crate::wa::session::PairMethod::Code,
        _ => crate::wa::session::PairMethod::Qr,
    };
    let pair_req = crate::wa::session::PairRequest {
        method,
        phone_e164: body.phone_e164.clone(),
    };
    let pair = crate::wa::pool::request_pair(&state, &session_id, pair_req).await?;

    let _ = audit::record(
        &state,
        AuditEntry {
            project_id: body.project_id.clone(),
            user_id: Some(body.user_id.clone()),
            session_id: Some(session_id.clone()),
            action: "session.pair".into(),
            target_kind: Some("session".into()),
            target_id: Some(session_id.clone()),
            metadata: serde_json::json!({
                "pairMethod": body.pair_method,
                "phoneE164": body.phone_e164,
            }),
            actor_ip,
            user_agent,
            ts: chrono::Utc::now(),
        },
    )
    .await;

    Ok(Json(CreateSessionResponse {
        session_id,
        status: "pending".into(),
        qr: pair.qr,
        pair_code: pair.pair_code,
    }))
}

async fn list_sessions(
    State(state): State<AppState>,
    Query(query): Query<ListSessionsQuery>,
) -> Result<Json<ListSessionsResponse>, AppError> {
    tracing::info!(project_id = %query.project_id, "sessions: list");

    let docs = crate::db::sessions::list_by_project(&state.db, &query.project_id).await?;
    let sessions = docs
        .into_iter()
        .map(|s| SessionSummary {
            session_id: s.id,
            project_id: s.project_id,
            user_id: s.user_id,
            phone_e164: s.phone_e164,
            status: s.status,
            push_name: s.push_name,
        })
        .collect();

    Ok(Json(ListSessionsResponse { sessions }))
}

async fn get_session(
    State(state): State<AppState>,
    Path(id): Path<String>,
) -> Result<Json<SessionSummary>, AppError> {
    tracing::info!(session_id = %id, "sessions: get");

    let s = crate::db::sessions::get(&state.db, &id).await?;
    Ok(Json(SessionSummary {
        session_id: s.id,
        project_id: s.project_id,
        user_id: s.user_id,
        phone_e164: s.phone_e164,
        status: s.status,
        push_name: s.push_name,
    }))
}

async fn delete_session(
    State(state): State<AppState>,
    headers: HeaderMap,
    Path(id): Path<String>,
) -> Result<Json<DeletedResponse>, AppError> {
    tracing::info!(session_id = %id, "sessions: delete");

    let (actor_ip, user_agent) = audit::extract_context(&headers);

    // TODO: provided by agent Rx — logout closes the live Baileys socket
    // before we drop persisted state.
    crate::wa::pool::logout(&state, &id).await?;
    crate::db::sessions::delete(&state.db, &id).await?;

    let _ = audit::record(
        &state,
        AuditEntry {
            project_id: String::new(),
            user_id: None,
            session_id: Some(id.clone()),
            action: "session.logout".into(),
            target_kind: Some("session".into()),
            target_id: Some(id.clone()),
            metadata: serde_json::json!({}),
            actor_ip,
            user_agent,
            ts: chrono::Utc::now(),
        },
    )
    .await;

    Ok(Json(DeletedResponse {
        session_id: id,
        deleted: true,
    }))
}

async fn update_session(
    State(state): State<AppState>,
    headers: HeaderMap,
    Path(id): Path<String>,
    Json(body): Json<UpdateSessionRequest>,
) -> Result<Json<UpdatedResponse>, AppError> {
    tracing::info!(session_id = %id, "sessions: update");

    let (actor_ip, user_agent) = audit::extract_context(&headers);

    crate::db::sessions::update(&state.db, &id, body.label.as_deref(), body.rate_limit_profile.as_deref()).await?;

    // Treat label-only changes as `session.rename` and anything else as a
    // generic `session.update`, mirroring the Next.js server-action shape.
    let action = if body.label.is_some() && body.rate_limit_profile.is_none() {
        "session.rename"
    } else {
        "session.update"
    };

    let _ = audit::record(
        &state,
        AuditEntry {
            project_id: String::new(),
            user_id: None,
            session_id: Some(id.clone()),
            action: action.into(),
            target_kind: Some("session".into()),
            target_id: Some(id.clone()),
            metadata: serde_json::json!({
                "label": body.label,
                "rateLimitProfile": body.rate_limit_profile,
            }),
            actor_ip,
            user_agent,
            ts: chrono::Utc::now(),
        },
    )
    .await;

    Ok(Json(UpdatedResponse {
        session_id: id,
        updated: true,
    }))
}

async fn get_session_status(
    State(state): State<AppState>,
    Path(id): Path<String>,
) -> Result<Json<SessionStatusResponse>, AppError> {
    tracing::info!(session_id = %id, "sessions: status");

    // TODO: provided by agent Rx — `status` reads from the live worker pool
    // (current QR / pair code if pending, otherwise connection state).
    let status = crate::wa::pool::status(&state, &id).await?;

    Ok(Json(SessionStatusResponse {
        session_id: id,
        status: status.status,
        qr: status.qr,
        pair_code: status.pair_code,
        last_connected_at: status.last_connected_at,
    }))
}

File: services/sabwa-node/src/routes/sessions.ts
/**
 * `/v1/sessions` — pairing, lifecycle and status for linked WhatsApp accounts.
 *
 * Wire shape matches the Rust engine's `src/routes/sessions.rs` so the
 * Next.js `engineFetch` client keeps working unchanged:
 *
 *   POST   /v1/sessions               → create + start pairing
 *   GET    /v1/sessions?projectId=    → list sessions for a project
 *   GET    /v1/sessions/:id           → single session summary
 *   PATCH  /v1/sessions/:id           → update label / rateLimitProfile
 *   DELETE /v1/sessions/:id           → logout + wipe row + evict from pool
 *
 * Every handler reads the shared `AppState` from `req.app.locals.state`.
 */

import { Router, type Request, type Response } from 'express';
import type { AppState } from '../state.js';
import {
  deleteSession,
  findById,
  findByProject,
  insertPending,
  updateMetadata,
  type PairMethod,
  type RateProfile,
  type SessionSummary,
} from '../db/sessions.js';
import { asString } from './_helpers.js';

/** Pull the request-scoped `AppState`. */
function getState(req: Request): AppState {
  const s = req.app.locals.state as AppState | undefined;
  if (!s) throw new Error('AppState missing from app.locals');
  return s;
}

/** Validate `pairMethod`. */
function parsePairMethod(raw: unknown): PairMethod | undefined {
  if (raw === 'qr' || raw === 'code') return raw;
  return undefined;
}

function parseRateProfile(raw: unknown): RateProfile | undefined {
  if (raw === 'safe' || raw === 'normal' || raw === 'aggressive') return raw;
  return undefined;
}

/**
 * Wait briefly for a QR string to appear after the socket spawns. Baileys
 * emits the QR shortly after `start()` resolves, so we poll the in-memory
 * cache rather than blocking on a one-shot event. Returns whichever
 * artifact is known first (qr or pairCode).
 */
async function waitForFirstArtifact(
  pool: AppState['pool'],
  sessionId: string,
  timeoutMs: number,
  preferPairCode: boolean,
  phoneE164: string | undefined,
): Promise<{ qr?: string; pairCode?: string }> {
  const deadline = Date.now() + timeoutMs;
  // Poll every 100ms — Baileys typically emits the first QR within ~1-2s.
  while (Date.now() < deadline) {
    const session = pool.get(sessionId);
    if (!session) return {};
    const status = session.getStatus();
    if (preferPairCode) {
      if (status.pairCode) return { pairCode: status.pairCode };
      // Once a QR has appeared the socket is ready to accept
      // `requestPairingCode`; trigger the request if it hasn't fired yet.
      if (status.qr && phoneE164) {
        const code = await session.getPairCode(phoneE164);
        if (code) return { pairCode: code };
      }
    } else if (status.qr) {
      return { qr: status.qr };
    }
    await new Promise((r) => setTimeout(r, 100));
  }
  return {};
}

// ----- Handlers -----

async function createSession(req: Request, res: Response): Promise<void> {
  const state = getState(req);
  const body = (req.body ?? {}) as Record<string, unknown>;
  const projectId = typeof body.projectId === 'string' ? body.projectId : '';
  const userId = typeof body.userId === 'string' ? body.userId : '';
  const pairMethod = parsePairMethod(body.pairMethod);
  const phoneE164 =
    typeof body.phoneE164 === 'string' && body.phoneE164.trim().length > 0
      ? body.phoneE164.trim()
      : undefined;

  if (!projectId) {
    res.status(400).json({ error: 'projectId is required', code: 'bad_request' });
    return;
  }
  if (!userId) {
    res.status(400).json({ error: 'userId is required', code: 'bad_request' });
    return;
  }
  if (!pairMethod) {
    res
      .status(400)
      .json({ error: "pairMethod must be 'qr' or 'code'", code: 'bad_request' });
    return;
  }
  if (pairMethod === 'code' && !phoneE164) {
    res.status(400).json({
      error: "pairMethod 'code' requires phoneE164",
      code: 'bad_request',
    });
    return;
  }

  // Insert a `pending` row first so `listSessions` reflects the new
  // account immediately and the post-pair persister can resolve it by id.
  let inserted: { sessionId: string };
  try {
    inserted = await insertPending(state.db, {
      projectId,
      userId,
      pairMethod,
      phoneE164,
    });
  } catch (err) {
    state.log.warn({ err }, 'sessions.insertPending failed');
    res.status(400).json({
      error: err instanceof Error ? err.message : 'failed to insert session',
      code: 'bad_request',
    });
    return;
  }

  // Spin up the live Baileys socket in the in-process pool.
  try {
    await state.pool.getOrCreate(inserted.sessionId, {
      projectId,
      pairMethod,
      phoneE164,
      db: state.db,
      redis: state.redis,
      authStateKey: state.authStateKey,
      log: state.log,
    });
  } catch (err) {
    state.log.warn(
      { err, sessionId: inserted.sessionId },
      'pool.getOrCreate failed — leaving pending row',
    );
  }

  // Best-effort: wait up to 5s for the first QR / pair code. The browser
  // also subscribes via SSE for the streamed updates, so returning empty
  // here is fine — it just means the user sees the spinner a moment longer.
  const artifacts = await waitForFirstArtifact(
    state.pool,
    inserted.sessionId,
    5_000,
    pairMethod === 'code',
    phoneE164,
  );

  res.status(200).json({
    sessionId: inserted.sessionId,
    status: 'pending',
    qr: artifacts.qr,
    pairCode: artifacts.pairCode,
  });
}

async function listSessions(req: Request, res: Response): Promise<void> {
  const state = getState(req);
  const projectId = asString(req.query.projectId);
  if (!projectId) {
    res
      .status(400)
      .json({ error: 'projectId query param is required', code: 'bad_request' });
    return;
  }
  const rows = await findByProject(state.db, projectId);
  res.json({ sessions: rows });
}

async function getSession(req: Request, res: Response): Promise<void> {
  const state = getState(req);
  const id = asString(req.params.id);
  const row = await findById(state.db, id);
  if (!row) {
    res.status(404).json({ error: 'session not found', code: 'not_found' });
    return;
  }
  res.json(row satisfies SessionSummary);
}

async function patchSession(req: Request, res: Response): Promise<void> {
  const state = getState(req);
  const id = asString(req.params.id);
  const body = (req.body ?? {}) as Record<string, unknown>;

  const label =
    typeof body.label === 'string' ? body.label : undefined;
  const rateLimitProfile = parseRateProfile(body.rateLimitProfile);
  const warmupEnabled = typeof body.warmupEnabled === 'boolean' ? body.warmupEnabled : undefined;
  const dailyResetTimezone = typeof body.dailyResetTimezone === 'string' ? body.dailyResetTimezone : undefined;
  const overrides = typeof body.overrides === 'object' && body.overrides !== null ? body.overrides as Record<string, number> : undefined;

  if (label === undefined && rateLimitProfile === undefined && warmupEnabled === undefined && dailyResetTimezone === undefined && overrides === undefined) {
    res.status(400).json({
      error: 'no patchable fields supplied',
      code: 'bad_request',
    });
    return;
  }

  await updateMetadata(state.db, id, { label, rateLimitProfile, warmupEnabled, dailyResetTimezone, overrides });
  res.json({ sessionId: id, updated: true });
}

/**
 * GET /v1/sessions/:id/status
 *
 * Returns a lightweight status snapshot for a single session.
 * Checks the live pool first (for `qr`, `pairCode`, `lastConnectedAt`,
 * `lastError`) and falls back to the persisted `sabwa_sessions` row for
 * sessions that are not currently held in the pool (e.g. after a restart).
 *
 * Shape mirrors `SabwaSessionStatusInfo` in `sabwa.actions.ts`.
 */
async function getSessionStatus(req: Request, res: Response): Promise<void> {
  const state = getState(req);
  const id = asString(req.params.id);
  if (!id) {
    res.status(400).json({ error: 'session id is required', code: 'bad_request' });
    return;
  }

  // Try the live pool first — it has the freshest ephemeral state (QR, pair code).
  const live = state.pool.get(id);
  if (live) {
    const snap = live.getStatus();
    res.json({
      session: {
        sessionId: id,
        status: snap.status,
        lastConnectedAt: snap.lastConnectedAt?.toISOString(),
        lastError: snap.lastError,
      },
    });
    return;
  }

  // Pool miss — look up the persisted row.
  const row = await findById(state.db, id);
  if (!row) {
    res.status(404).json({ error: 'session not found', code: 'not_found' });
    return;
  }
  res.json({
    session: {
      sessionId: row.sessionId,
      status: row.status,
      phoneE164: row.phoneE164,
      pushName: row.pushName,
      profilePicUrl: row.profilePicUrl,
      lastConnectedAt: row.lastConnectedAt,
    },
  });
}

async function getRateLimitProfile(req: Request, res: Response): Promise<void> {
  const state = getState(req);
  const id = asString(req.params.id);
  const row = await findById(state.db, id);
  if (!row) {
    res.status(404).json({ error: 'session not found', code: 'not_found' });
    return;
  }
  
  // Note: findById uses toSummary which doesn't include the new rate limit fields,
  // we need to fetch the raw doc to return them.
  const { ObjectId } = await import('mongodb');
  let oid;
  try { oid = new ObjectId(id); } catch { res.status(404).json({ error: 'session not found', code: 'not_found' }); return; }
  const doc = await state.db.collection('sabwa_sessions').findOne({ _id: oid });
  if (!doc) {
    res.status(404).json({ error: 'session not found', code: 'not_found' });
    return;
  }

  const createdAt = doc.createdAt;
  const sessionAgeDays = createdAt ? Math.floor((Date.now() - createdAt.getTime()) / (1000 * 60 * 60 * 24)) : 0;

  res.json({
    settings: {
      profile: doc.rateLimitProfile || 'normal',
      warmupEnabled: doc.warmupEnabled ?? false,
      dailyResetTimezone: doc.dailyResetTimezone ?? 'UTC',
      overrides: doc.overrides ?? {},
      sessionAgeDays,
    }
  });
}

async function deleteSessionHandler(
  req: Request,
  res: Response,
): Promise<void> {
  const state = getState(req);
  const id = asString(req.params.id);

  // Logout the live socket (best-effort), then evict from the pool, then
  // wipe the row + auth_state. The pool eviction must happen before the
  // delete so the post-eviction reconnect path can't race against the
  // wipe and re-insert an empty `creds.update`.
  const live = state.pool.remove(id);
  if (live) {
    try {
      await live.logout();
    } catch (err) {
      state.log.warn(
        { err, sessionId: id },
        'live session logout failed — continuing with row delete',
      );
    }
  }
  // `logout()` already deletes the Mongo row, but call again in case the
  // session wasn't in the pool (e.g. server restarted after pairing).
  await deleteSession(state.db, id);
  res.json({ sessionId: id, deleted: true });
}

// ----- Router -----

/** Build the `/v1/sessions` sub-router. */
export function buildSessionsRouter(_state: AppState): Router {
  const r = Router();
  r.post('/', (req, res, next) => {
    createSession(req, res).catch(next);
  });
  r.get('/', (req, res, next) => {
    listSessions(req, res).catch(next);
  });
  // Static sub-paths must be registered before the dynamic /:id handler
  // so Express doesn't swallow 'status' / 'rate-limits' as the :id segment.
  r.get('/:id/status', (req, res, next) => {
    getSessionStatus(req, res).catch(next);
  });
  r.get('/:id/rate-limits', (req, res, next) => {
    getRateLimitProfile(req, res).catch(next);
  });
  r.get('/:id', (req, res, next) => {
    getSession(req, res).catch(next);
  });
  r.patch('/:id', (req, res, next) => {
    patchSession(req, res).catch(next);
  });
  r.delete('/:id', (req, res, next) => {
    deleteSessionHandler(req, res).catch(next);
  });
  return r;
}
